compiler/noirc_frontend/src/elaborator/mod.rs

```diff
@@ -733,24 +734,98 @@ impl<'context> Elaborator<'context> {
-    /// there's a bunch of other places where trait constraints can pop up
     fn resolve_trait_constraints(
         &mut self,
         where_clause: &[UnresolvedTraitConstraint],
+        add_constraints_to_scope: bool,
     ) -> Vec<TraitConstraint> {
         where_clause
             .iter()
-            .filter_map(|constraint| self.resolve_trait_constraint(constraint))
+            .filter_map(|constraint| {
+                self.resolve_trait_constraint(constraint, add_constraints_to_scope)
+            })
+            .collect()
+    }
+
+    /// Expands any traits in a where clause to mention all associated types if they were
+    /// ellided by the user. See `add_missing_named_generics` for more  detail.
+    ///
+    /// Returns all newly created generics to be added to this function/trait/impl.
+    fn desugar_trait_constraints(
+        &mut self,
+        where_clause: &mut [UnresolvedTraitConstraint],
+    ) -> Vec<ResolvedGeneric> {
+        where_clause
+            .iter_mut()
+            .flat_map(|constraint| self.add_missing_named_generics(&mut constraint.trait_bound))
             .collect()
     }
 
-    pub fn resolve_trait_constraint(
+    /// For each associated type that isn't mentioned in a trait bound, this adds
+    /// the type as an implicit generic to the where clause and returns the newly
+    /// created generics in a vector to add to the function/trait/impl later.
+    /// For example, this will turn a function using a trait with 2 associated types:
+    ///
+    /// `fn foo<T>() where T: Foo { ... }`
+    ///
+    /// into:
+    /// `fn foo<T>() where T: Foo<Bar = A, Baz = B> { ... }`
+    ///
+    /// with a vector of `<A, B>` returned so that the caller can then modify the function to:
+    /// `fn foo<T, A, B>() where T: Foo<Bar = A, Baz = B> { ... }`
+    fn add_missing_named_generics(&mut self, bound: &mut TraitBound) -> Vec<ResolvedGeneric> {
+        let mut added_generics = Vec::new();
+
+        let Some(the_trait) = self.lookup_trait_or_error(bound.trait_path.clone()) else {
+            return added_generics;
+        };
+
+        if the_trait.associated_types.len() > bound.trait_generics.named_args.len() {
+            for associated_type in the_trait.associated_types.clone() {
+                if !bound
+                    .trait_generics
+                    .named_args
+                    .iter()
+                    .any(|(name, _)| name.0.contents == *associated_type.name.as_ref())
+                {
+                    // This generic isn't contained in the bound's named arguments,
+                    // so add it by creating a fresh type variable.
+                    let new_generic_id = self.interner.next_type_variable_id();
+                    let type_var = TypeVariable::unbound(new_generic_id, Kind::Normal);
+
+                    let span = bound.trait_path.span;
+                    let no_name = Rc::new("_".to_string());
```

Comment:
I'd prefer for this to be more distinct from a potentially user-entered label if possible, e.g.
```suggestion
                    let name = Rc::new(&format("_{}", name.0.contents));
```

Reply:
I thought showing `T: Foo<Bar = _>` was more readable to the user than `T: Foo<Bar = Bar>` or `T: Foo<Bar = _Bar>` since neither `Bar` nor `_Bar` are actual types in scope.

I haven't tested it much but if this type leaks into other errors I could see it being easier to read if we used `Bar` instead. E.g. if you get an error for `No impl Eq for _` versus `No impl Eq for Bar`. Haven't tested how easy that is to trigger though.

Reply:
I was able to trigger this with
```noir
fn baz<T>(x: T) where T: Foo {
    let y = x.foo();
    assert_eq(y, y);
}
```
"No matching impl for _: Eq"
so I changed it to use the associated type's name instead which should hopefully be more clear where the type is coming from. It also means users will see something like `Foo<Bar = Bar>` often, but we can try to change the display in the future to leave that out somehow.